apps/conv_layer/conv_layer_generator.cpp

```diff
@@ -10,72 +10,185 @@ class ConvolutionLayer : public Halide::Generator<ConvolutionLayer> {
     Input<Buffer<float>> filter{"filter", 4};
     Input<Buffer<float>> bias{"bias", 1};
 
-    Output<Buffer<float>> f_ReLU{"ReLU", 4};
+    Output<Buffer<float>> relu{"relu", 4};
 
     void generate() {
-        /* THE ALGORITHM */
+        const int N = 5, CI = 128, CO = 128, W = 100, H = 80;
 
-        Var x("x"), y("y"), z("z"), n("n");
+        /* THE ALGORITHM */
 
-        Func f_conv("conv");
-        RDom r(filter.dim(0).min(), filter.dim(0).extent(),
-               filter.dim(1).min(), filter.dim(1).extent(),
-               filter.dim(2).min(), filter.dim(2).extent());
+        Var x("x"), y("y"), c("c"), n("n");
 
-        f_conv(x, y, z, n) = bias(z);
+        Func conv("conv");
+        RDom r(0, CI, 0, 3, 0, 3);
 
-        f_conv(x, y, z, n) += filter(r.x, r.y, r.z, z) * input(x + r.x, y + r.y, r.z, n);
+        conv(c, x, y, n) = bias(c);
+        conv(c, x, y, n) += filter(c, r.y, r.z, r.x) * input(r.x, x + r.y, y + r.z, n);
 
-        f_ReLU(x, y, z, n) = max(0, f_conv(x, y, z, n));
+        relu(c, x, y, n) = max(0, conv(c, x, y, n));
 
         /* THE SCHEDULE */
 
+        // MKL JITs code for the specific size and strides, so we'll
+        // do the same and ask Halide to compile for this specific
+        // size:
+        relu.bound(x, 0, W)
+            .bound(y, 0, H)
+            .bound(c, 0, CO)
+            .bound(n, 0, N);
+
+        relu.dim(0).set_bounds(0, CO).set_stride(1);
+        relu.dim(1).set_bounds(0, W).set_stride(CO);
+        relu.dim(2).set_bounds(0, H).set_stride(CO * W);
+        relu.dim(3).set_bounds(0, N).set_stride(CO * H * W);
+
+        input.dim(0).set_bounds(0, CI).set_stride(1);
+        input.dim(1).set_bounds(0, W + 2).set_stride(CI);
+        input.dim(2).set_bounds(0, H + 2).set_stride(CI * (W + 2));
+        input.dim(3).set_bounds(0, N).set_stride(CI * (W + 2) * (H + 2));
+
+        filter.dim(0).set_bounds(0, CO).set_stride(1);
+        filter.dim(1).set_bounds(0, 3).set_stride(CO);
+        filter.dim(2).set_bounds(0, 3).set_stride(CO * 3);
+        filter.dim(3).set_bounds(0, CI).set_stride(CO * 3 * 3);
+
+        bias.dim(0).set_bounds(0, CO).set_stride(1);
+
+        input.set_host_alignment(128);
```

Comment:
I'm curious, this is new in this PR (so not a remanent of a Hexagon schedule?), did this have some impact on  CPU/GPU performance?

Reply:
I was just copying over the code from the autoscheduler branch. I'm not sure how important it is, but it certainly makes it easier to understand the generated code. Probably there are some alignment-related benefits. (CUDA only does vector loads from aligned addresses)

Reply:
Oh yeah, if you're referring specifically to the host alignment constraints, it's so that pairs of 32-bit loads can be done as a single 64-bit load instead, to reduce the number of memory transactions. 128 is probably overkill.

Reply:
Yeah, I was wondering specifically about set_host_alignment. It hadn't occurred to me that it might significantly affect CPU codegen. Maybe I should try it on more code...

Reply:
I don't think it matters for modern CPUs - the memory transaction minimizing trick is just for CUDA.

Reply:
OK that makes more sense. I was confused about this but now it's clear.

I think the alignment requirement should be reduced to what is necessary, and moved inside the GPU specific schedule.

Reply:
Turned out to be unnecessary (still not entirely sure why), so I just deleted the alignment constraints.